src/core/geometry/qgscompoundcurve.cpp

```diff
@@ -906,6 +907,101 @@ QVector< QPair<int, QgsVertexId> > QgsCompoundCurve::curveVertexId( QgsVertexId
   return curveIds;
 }
 
+bool QgsCompoundCurve::convertVertex( QgsVertexId position )
+{
+
+  // First we find out the sub-curves that are contain that vertex.
+
+  // If there is more than one, it means the vertex was at the beginning or end
+  // of an arc, which we don't support.
+
+  // If there is exactly one, we may either be on a LineString, or on a CircularString.
+
+  // If on CircularString, we need to check if the vertex is a CurveVertex (odd index).
+  // If so, we split the subcurve at vertex -1 and +1, , drop the middle part and insert a LineString/CircularString
+  // instead with the same points.
+  
+  // At the end, we call condenseCurves() to merge successible line/circular strings
+
+  QVector< QPair<int, QgsVertexId> > curveIds = curveVertexId( position );
+
+  // We cannot convert points at start/end of subcurves
+  if ( curveIds.length() != 1 )
+    return false;
+
+  int curveId = curveIds[0].first;
+  QgsVertexId subVertexId = curveIds[0].second;
+  QgsCurve *curve = mCurves[curveId];
+
+  // We cannot convert first/last point of curve
+  if ( subVertexId.vertex == 0 || subVertexId.vertex == curve->numPoints() - 1 )
+    return false;
+
+  // TODO factorize circularString and lineString as logic is almost the same
+  if ( const QgsCircularString *circularString = dynamic_cast<const QgsCircularString *>( curve ) )
+  {
+    // If it's a circular string, we convert to LineString
+
+    // We cannot convert start/end points of arcs
+    if ( subVertexId.vertex % 2 == 0 ) // for some reason, subVertexId.type is always SegmentVertex...
+      return false;
+
+    QgsPointSequence points;
+    circularString->points( points );
+
+    QgsPointSequence partA  = points.mid( 0, subVertexId.vertex );
+    QgsPointSequence partB  = QgsPointSequence() << points[subVertexId.vertex - 1] << points[subVertexId.vertex] << points[subVertexId.vertex + 1];
+    QgsPointSequence partC  = points.mid( subVertexId.vertex + 1 );
+
+    QgsCircularString *curveA = new QgsCircularString();
+    curveA->setPoints( partA );
+    QgsLineString *curveB = new QgsLineString();
+    curveB->setPoints( partB );
+    QgsCircularString *curveC = new QgsCircularString();
+    curveC->setPoints( partC );
+
+    removeCurve( curveId );
+    if ( subVertexId.vertex < points.length() - 2 )
+      mCurves.insert( curveId, curveC );
+    mCurves.insert( curveId, curveB );
+    if ( subVertexId.vertex > 1 )
+      mCurves.insert( curveId, curveA );
```

Comment:
```suggestion
    std::unique_ptr<QgsCircularString> curveA = std::make_unique<QgsCircularString>();
    curveA->setPoints( partA );
    std::unique_ptr<QgsLineString> curveB = std::make_unique<QgsLineString>();
    curveB->setPoints( partB );
    std::unique_ptr<QgsCircularString> curveC = std::make_unique<QgsCircularString>();
    curveC->setPoints( partC );

    removeCurve( curveId );
    if ( subVertexId.vertex < points.length() - 2 )
      mCurves.insert( curveId.release(), curveC.release() );
    mCurves.insert( curveId, curveB.release() );
    if ( subVertexId.vertex > 1 )
      mCurves.insert( curveId, curveA.release() );
```

This should avoid leaking memory here